[EPPETA-10] EvaluationDbContext, DI, Model

StevenM0 opened the pull request.

StevenM0 committed:
EvaluationDbContext, DI, Model

stephenfuqua commented on src/webapi/Evaluations/Models/PerformanceEvaluationRating.cs:
Will the SQL table that matches this class be added in a pull request for EPPETA-11?

stephenfuqua commented on src/webapi/Evaluations/Data/EvaluationDbContext.cs:
Entity Framework has a concept of entity tracking, which is not useful in web applications (easier to explain in person than in writing here). I forgot to disable entity  tracking when I was working on the security layer. Please see add the code [from this example](https://learn.microsoft.com/en-us/ef/core/querying/tracking#configuring-the-default-tracking-behavior) to Program.cs for the IdentityDbContext, and you'll need to add a dependency injection step for this new context as well.

stephenfuqua replied:
I see the dependency injection is done now. After talking through the Entity Framework tracking issue, we decided not to make a change with the tracking, because there may be implications that I'm forgetting about when trying to update an existing record. Don't want to overly complicate things with this type of pre-optimization.

StevenM0 committed:
evaluation sql, rating model classes, DbSets, FK

stephenfuqua commented on sql/mssql/00020-Evaluation-Tables.sql:
```diff
@@ -0,0 +1,149 @@
+    PerformanceEvaluationDescription VARCHAR(255) NULL,
+    CreateDate DATETIME2 NOT NULL,
+    LastModifiedDate DATETIME2 NOT NULL,
+    Ods_Id UNIQUEIDENTIFIER NULL,
```
Suggestion: rename to `EdFi_Id` and change to a string. "ODS" is the name of a database, not an API. So we want to make this a little more generic. Normally we don't use underscores in SQL Server naming conventions, but I like your use of it here, especially when changing the name: `EdFiId` is hard to read 😆 .

Also change to a string so that we can accommodate other versions of an Ed-Fi API that use strings but not necessarily UUID values. `varchar(50)` seems fine.

```suggestion
    EdFi_Id VARCHAR(50) NULL,
```

stephenfuqua commented on sql/mssql/00020-Evaluation-Tables.sql:
```diff
@@ -0,0 +1,149 @@
+CREATE TABLE eppeta.PerformanceEvaluation (
+    EducationOrganizationId INT NOT NULL,
+    EvaluationPeriodDescriptorId INT NOT NULL,
+    PerformanceEvaluationTitle VARCHAR(50) NOT NULL,
```
I agree with use of `VARCHAR`. Just an opportunity to comment: any database that needs to support internationalization would need `NVARCHAR` for support of UTF-16. Since Ed-Fi is only for the US market, that kind of internationalization for support of highly extended character sets seems very unlikely, so `VARCHAR` is safe for now.

stevenarnold-ed-fi replied:
Not sure how these ended up as varchar, but they should be nvarchar, @StevenM0 please update any VARCHAR to NVARCHAR

stephenfuqua replied:
@stevenarnold-ed-fi why would we use NVARCHAR here? I don't mean to make a big deal about it, but I don't see a compelling reason for it.

stephenfuqua replied:
@stevenarnold-ed-fi do you want to update to `NVARCHAR` in this PR?

stevenarnold-ed-fi replied:
Updated for consistency with authentication tables

stephenfuqua commented on sql/mssql/00020-Evaluation-Tables.sql:
```diff
@@ -0,0 +1,149 @@
+    CreateDate DATETIME2 NOT NULL,
+    LastModifiedDate DATETIME2 NOT NULL,
+    Ods_Id UNIQUEIDENTIFIER NULL,
+    Id INT IDENTITY(1,1) PRIMARY KEY,
```
I like to be explicit about naming things.
```suggestion
    Id INT IDENTITY(1,1) NOT NULL,
    CONSTRAINT PerformanceEvaluation_PK PRIMARY KEY CLUSTERED (Id ASC)
```

stephenfuqua commented on sql/mssql/00020-Evaluation-Tables.sql:
```diff
@@ -0,0 +1,149 @@
+    SchoolYear SMALLINT NOT NULL,
+    TermDescriptorId INT NOT NULL,
+    PerformanceEvaluationDescription VARCHAR(255) NULL,
+    CreateDate DATETIME2 NOT NULL,
+    LastModifiedDate DATETIME2 NOT NULL,
```
`datetime2` has more precision - thus higher storage requirement - than `datetime`. We don't need that level of precision here.
```suggestion
    CreateDate DATETIME NOT NULL,
    LastModifiedDate DATETIME NOT NULL,
```

stephenfuqua commented on sql/mssql/00020-Evaluation-Tables.sql:
```diff
@@ -0,0 +1,149 @@
+    ActualDate DATE NOT NULL,
+    ActualDuration INT NULL,
+    PerformanceEvaluationRatingLevelDescriptorId INT NULL,
+    ActualTime TIME NULL,
```
I have been using SQL Server continuously since 2005 and I did not know there is a `TIME` data type. I have literally never seen it be used before 😄. It might be better to just leave this as a string or a numeric type, depending on how you are going to use it in C#.

stevenarnold-ed-fi replied:
![image](https://github.com/Ed-Fi-Exchange-OSS/EvaluationTracker/assets/48261255/05555632-0762-4afb-a404-0985ee29832d)
This will be a user entered field that might be used to calculate duration within this app and then transferred through a string field on the API and will land in a TIME field within the ODS. I took the ODS SQL and trimmed it down to get these tables. I have no real opinion on TIME vs another datatype here

stephenfuqua replied:
If the value is already a `TIME` data type in the ODS then sounds fine to me to keep it that way.

stephenfuqua commented on src/webapi/Evaluations/Models/EvaluationElementRating.cs:
```diff
@@ -0,0 +1,29 @@
+        public decimal Rating { get; set; }
+        public DateTime CreateDate { get; set; }
+        public DateTime LastModifiedDate { get; set; }
+        public Guid? Ods_Id { get; set; }
```
```suggestion
        public string? EdFi_Id { get; set; }
```

stephenfuqua reviewed (commented):
These tables are well designed. Just some comments on fine details.

StevenM0 committed:
EvaluationRatingController, Define many-to-one relationship, revised SQL scripts and model classes

StevenM0 committed:
retrieve ratings and modify landing page

stevenarnold-ed-fi committed:
Retrieve PerformanceEvaluationRatings by logged in user + general cleanup

- added jwt-decode package for decoding of JWT on front end
- update JWT storage/retrieval to localStorage
- updated landing page to retrieve only logged in users evaluations
- parameterize some ODS/API URL's in C# code
- parameterize ODS/API key/secret
- introduced PerformanceEvaluationRating DTO to handle evaluator name to front end
- Introduced mapping helper class and PerformanceEvaluationRating entity to DTO mapping
- disabled OpenIdDict JWT encryption

stevenarnold-ed-fi committed:
Updating so API URI will flow from appsettings to authentication configuration again

stephenfuqua commented on src/package-lock.json:
We discussed removing this and `package.json` in `src` folder.

stevenarnold-ed-fi replied:
Removed

stevenarnold-ed-fi committed:
Updates from PR
